### DISCUSSION TITLE
Usage of tuples/composite types in functions like `count`

### DISCUSSION THREAD

**Discussion by oeed** [diesel-rs/diesel#4309] [Ideas]
_An issue/discussion might already exist about this, although I wasn't able to find one so sorry if it does (I seem to recall discussing this at some point, but maybe it was on Gitter)._

I have a situation where I need to count distinct values in a column that is a composite type/tuple (Postgres). For a highly contrived example:

```sql
CREATE TYPE person_info AS (
    name TEXT,
    age INT
);

CREATE TABLE employees (
    id SERIAL PRIMARY KEY,
    info person_info
);

SELECT COUNT(DISTINCT info) FROM employees;
```

The current count functions (along with other similar functions) restrict the expression to `SingleValue`, however, which isn't implemented on tuples. Is there perhaps another more suitable that these bounds could be changed to, or perhaps a new trait for such values on the Postgres backend? Thoughts?

**Comment by weiznich** [comment]
Given your example I'm not sure I can follow. There is already support for [`COUNT(DISTINCT …)`](https://docs.diesel.rs/2.2.x/diesel/dsl/fn.count_distinct.html), so your example query should be just `employees::table.select(dsl::count_distinct(employees::info))`

Or do you look for passing more than one value to distinct? In that case: Yes that's not supported yet. The situation there is as with all the other issues/request to allow tuples there: It needs someone that work on a design first. We are open to suggestions there.

**Reply by oeed** [reply]
Ah apologies, I see. The issue I was making was trying to use a pure tuple was my column SQL type. I instead needed to use the `Record` type to wrap my tuple, which does implement `SingleValue`. Thanks
